sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -190,7 +191,36 @@ case class AdaptiveSparkPlanExec(
           executionId.foreach(onUpdatePlan(_, result.newStages.map(_.plan)))
 
           // Start materialization of all new stages and fail fast if any stages failed eagerly
-          result.newStages.foreach { stage =>
+
+          // SPARK-33933: we should materialize broadcast stages first and wait the
+          // materialization finish before materialize other stages, to avoid waiting
+          // for broadcast tasks to be scheduled and leading to broadcast timeout.
+          val broadcastMaterializationFutures = result.newStages
+            .filter(_.isInstanceOf[BroadcastQueryStageExec])
+            .map { stage =>
+            var future: Future[Any] = null
+            try {
+              future = stage.materialize()
+              future.onComplete { res =>
+                if (res.isSuccess) {
+                  events.offer(StageSuccess(stage, res.get))
+                } else {
+                  events.offer(StageFailure(stage, res.failed.get))
+                }
+              }(AdaptiveSparkPlanExec.executionContext)
+            } catch {
+              case e: Throwable =>
+                cleanUpAndThrowException(Seq(e), Some(stage.id))
+            }
+            future
+          }
+
+          // Wait for the materialization of all broadcast stages finish
```

Comment:
> In non-AQE, we always wait the broadcast finish before submit shuffle map tasks.

Is that true?

I think we submit broadcast task when preparing a SparkPlan, but I don't think we wait for it finished before running on other tasks.

Reply:
That's true, I also verified about it. You can see the implementation of BroadcastHashJoinExec. codegen will call prepareRelation() -> prepareBroadcast() and then call buildPlan.executeBroadcast[HashedRelation](), finally BroadcastExchangeExec.doExecuteBroadcast() will be called, which will get relationFuture and wait the broadcast finish.

Reply:
@cloud-fan please correct me if I am wrong. I am not sure when root node is not WholeStageCodegen.

Reply:
Once we are going to execute a query, first thing is to prepare the SparkPlan and broadcast is triggered at preparing. We don't wait it and continue the execution until we reach the place we really need broadcasted result. At the place we call `executeBroadcast` which waits the result of broadcast.

So you can see we don't wait after the broadcast task is triggered.

Reply:
I think we talk about the same thing. For non-AQE, usually it's one job for one query. So when we execute the query, transform it to RDD, we will need the broadcasted result. This is the place I mentioned. Before submit the job of the query, we will waits the result of broadcast.

here's part of the log I printed:
`
1610935584070 [Thread[ScalaTest-run-running-AdaptiveQueryExecSuite,5,main]] class org.apache.spark.sql.execution.WholeStageCodegenExec SparkPlan.prepare 
1610935584072 BroadcastExchangeExec.doPrepare()
1610935584073 beforeCollect 
1610935584074 [Thread[broadcast-exchange-0,5,main]] class org.apache.spark.sql.execution.CoalesceExec SparkPlan.prepare 
1610935584163 sc.runJob 
1610935584223 DAGScheduler.runJob 
1610935584224 DAGScheduler.submitJob Id = 0 
1610935584225 eventProcessLoop.post JobSubmitted 
1610935597943 beforeBuild 
1610935597982 promise.trySuccess(broadcasted) 
1610935597988 [Thread[ScalaTest-run-running-AdaptiveQueryExecSuite,5,main]] class org.apache.spark.sql.execution.InputAdapter SparkPlan.prepare 
1610935598184 sc.runJob 
1610935598194 DAGScheduler.runJob 
1610935598195 DAGScheduler.submitJob Id = 1 
1610935598195 eventProcessLoop.post JobSubmitted 
`

Reply:
No. For normal query without AQE, broadcast job is triggered in preparing the SparkPlan (please see `BroadcastExchangeExec.relationFuture`) and it is different to the job of the query itself. Then we _don't_ wait here. This is the main difference to your change here. Immediately after triggering the broadcast task, Spark continues other part of the query, until it _really_ needs the broadcasted result (please see `executeBroadcast`).

But here you wait for the materialization of broadcast task finishes. So even there are still resources available to run other shuffle stages, they won't be run. It is different to current AQE and non-AQE query execution.

Reply:
For normal query without AQE, I mean the query except the broadcast part is one job. So the job must submit to DAGScheduler after the broadcast. If the shuffle stage can be submitted before the broadcast finish, how to explain the log above for the UT?

Reply:
It is submitted after the broadcast, I don't say it is submitted _before_ the broadcast. I said, the broadcast is triggered during preparing the SparkPlan, but the query won't stop and wait for it to finish...

The point is, whether the shuffle stage needs to _wait for_ the broadcast finishes.

Reply:
> the broadcast is triggered during preparing the SparkPlan
Sorry for didn't make it clear. I agree with this from the beginning.

> The point is, whether the shuffle stage needs to wait for the broadcast finishes.
I think _needs to wait for_.  You can see from the logs above, the job 1 is submitted after the broadcast finish ( I print the log in relationFuture => promise.trySuccess ). I add sleep 2 sec in broadcast to make it clear.

Reply:
No, it maybe true for your run. In your case, maybe all resources are occupied by the broadcast. So no other job can be really scheduled to executors. But logically we don't set this limitation. I repeat, if there are enough resources, broadcast job and shuffle stage can be run in parallel.

Let me show it more clear...

1. During preparing the SparkPlan, broadcast job is submitted to run, right? And we don't stop and wait here.
2. Now Spark continues the execution of the query. If there is shuffle stage independent to the broadcast, it can be scheduled to run in parallel if there are enough resources in the cluster.
3. Only if Spark calls `executeBroadcast` of the broadcast's query plan, we really stop and wait for the broadcast result.

Is it clear to you?

Reply:
> No, it maybe true for your run. In your case, maybe all resources are occupied by the broadcast. So no other job can be really scheduled to executors. But logically we don't set this limitation. I repeat, if there are enough resources, broadcast job and shuffle stage can be run in parallel.

I can ensure there's enough resources to run broadcast and shuffle in parallel. Actually, there haven't come to TaskScheduler, "sc.runJob" is called after the broadcast finished.

> 
1. During preparing the SparkPlan, broadcast job is submitted to run, right? And we don't stop and wait here.
2. Now Spark continues the execution of the query. If there is shuffle stage independent to the broadcast, it can be scheduled to run in parallel if there are enough resources in the cluster.
3. Only if Spark calls executeBroadcast of the broadcast's query plan, we really stop and wait for the broadcast result.

I agree with you for 1 and 3, but cannot agree with you for 2. In non-AQE, break a job into stages is happened in DAGScheduler, it is after physical plan has been transformed to RDD. I don't think DAGScheduler can call the executeBroadcast, which is part of spark-sql.

Below is the physical plan.

== Physical Plan ==
* Project (13)
+- * BroadcastHashJoin Inner BuildRight (12)
   :- * HashAggregate (6)
   :  +- Exchange (5)
   :     +- * HashAggregate (4)
   :        +- * Project (3)
   :           +- * SerializeFromObject (2)
   :              +- Scan (1)
   +- BroadcastExchange (11)
      +- Coalesce (10)
         +- * Project (9)
            +- * SerializeFromObject (8)
               +- Scan (7)

You can see the submission time in the screenshot.

![image](https://user-images.githubusercontent.com/3882710/105017437-ba207300-5a7e-11eb-9365-5fee24f27ea7.png)

![image](https://user-images.githubusercontent.com/3882710/105017527-d58b7e00-5a7e-11eb-83f7-41babce76efc.png)

There are 4 cores and only 1 task for broadcast job:
![image](https://user-images.githubusercontent.com/3882710/105017619-f653d380-5a7e-11eb-9e57-9fc92d458a31.png)

Reply:
Ok, I see your point.

> I can ensure there's enough resources to run broadcast and shuffle in parallel. Actually, there haven't come to TaskScheduler, "sc.runJob" is called after the broadcast finished.

In short, `submitMapState` can be used to submit a job too. Although seems it is only called now in AQE. I'm surprised that we don't do it similarly in normal query.

> In non-AQE, we always wait the broadcast finish before submit shuffle map tasks.

I think this isn't true because we don't forcibly wait broadcast to be finished like the change here. Technically, Spark doesn't restrict a shuffle to be run in parallel with a broadcast. An operator can call `submitMapState` to trigger a shuffle map stage if it wants, and the map stage can be run in parallel with broadcast.

You could say in non-AQE currently shuffle map tasks are submitted after the broadcast finishes. But not we always wait for the broadcast.

Reply:
> In short, submitMapState can be used to submit a job too. Although seems it is only called now in AQE. I'm surprised that we don't do it similarly in normal query.

submitMapState in DAGScheduler is introduced in SPARK-9851 just for AQE propose. I also think it can be used in normal query. I guess the reason we don't is there is less benefit as the complex it introduced. The architecture of sql execution should be changed just like AQE, one job will be break to many job, etc..

> You could say in non-AQE currently shuffle map tasks are submitted after the broadcast finishes. But not we always wait for the broadcast.

It seems like the same thing. But we can make it better in AQE. If we don't like the solution, we can use the initial one  as partial fix and find a good solution, as discussed with @cloud-fan .

Reply:
I don't think we should block shuffle stages manually and wait for broadcast stages. The reason is like what @cloud-fan already mentioned. The first change looks better although it is not perfect too.